dill/_dill.py

```diff
@@ -410,9 +416,69 @@ def save_numpy_array(pickler, obj):
             raise PicklingError(msg)
         logger.trace_setup(self)
         StockPickler.dump(self, obj)
-
     dump.__doc__ = StockPickler.dump.__doc__
 
+    def save(self, obj, save_persistent_id=True, *, name=None):
+        """If self._refonfail is True, try to save object by reference if pickling fails."""
+        if not self._refonfail:
+            super().save(obj, save_persistent_id)
+            return
+        # Disable framing (right after the framer.init_framing() call at dump()).
+        self.framer.current_frame = None
+        # Store initial state.
+        position = self._file_tell()
+        memo_size = len(self.memo)
+        try:
+            super().save(obj, save_persistent_id)
+        except UNPICKLEABLE_ERRORS + (AttributeError,) as error_stack:
+            # AttributeError may happen in the save_global() call from a child object.
+            if (type(error_stack) == AttributeError
+                    and "no attribute '__name__'" not in error_stack.args[0]):
+                raise
+            # Roll back the stream.
+            self._file_seek(position)
+            self._file_truncate()
+            # Roll back memo.
+            for _ in range(len(self.memo) - memo_size):
+                self.memo.popitem()  # LIFO order is guaranteed since 3.7
```

Comment:
Have you tested this in the context of a `multiprocess.Pool` or a `multiprocess.dummy.Pool`?

Reply:
I can't even imagine how a pickler would be parallelized by using a `Pool`. The `memo` is an instance variable, and `memoize()` is not atomic. Calling `dump()` multiple times in parallel will probably fail (or generate corrupted pickles) even without these `refonfail`-related operations.

Reply:
something like this is a simple case of `dump` in parallel...
```
Python 3.7.13 (default, May 10 2022, 11:13:40) 
[Clang 10.0.1 (clang-1001.0.46.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 
>>> import multiprocess.dummy as threads
>>> p = threads.Pool()
>>> import dill
>>> y = p.map(dill.dumps, range(4))
>>> y
[b'\x80\x03K\x00.', b'\x80\x03K\x01.', b'\x80\x03K\x02.', b'\x80\x03K\x03.']
>>> 
```

Reply:
Well, in this example a new instance of `dill.Pickler` is created for each call to `dump()`. If it's this kind of thing you were concerned about, we are fine.

Reply:
It would be worthwhile to run the `multiprocess` test suite to see if there are any new failures due to the change.  Typically, when something significant is touched, it's a good idea to run (at least) the `multiprocess`, `pathos`, `klepto`, and `mystic` test suites. They are a good stress test for `dill`, as they use `dill` as intended but have some "advanced pickling" cases.

Reply:
I run the test suites you suggested and `pathos`, `klepto` and `mystic` passed without errors. I've got an error with `multiprocess`, but it is identical to what I get with the `master` branch (something related to resources).